src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.cpp

```diff
@@ -0,0 +1,298 @@
+
+        // robot will try to retreat backwards from wherever it is currently facing
+        Angle final_orientation  = nearest_robot.value().orientation();
+        Vector retreat_direction = (nearest_robot->position() - ball_pos).normalize();
+        Point retreat_position =
+            ball_pos + retreat_direction * (0.5 + ROBOT_MAX_RADIUS_METERS);
+
+        // if the initial retreat position is out of the field boundary, have it retreat
+        // towards the closest goal
+        if (!contains(world.field().fieldBoundary(), retreat_position))
+        {
+            bool in_friendly_half = contains(world.field().friendlyHalf(), ball_pos);
+            Point closer_goal     = in_friendly_half ? world.field().friendlyGoalCenter()
+                                                 : world.field().enemyGoalCenter();
+            retreat_direction = (closer_goal - ball_pos).normalize();
+            retreat_position =
+                ball_pos + retreat_direction * (0.5 + ROBOT_MAX_RADIUS_METERS);
+        }
+
+        // setup ball placement tactic for ball placing robot
+        retreat_tactic->updateControlParams(
+            retreat_position, final_orientation, 0.0, TbotsProto::DribblerMode::OFF,
+            TbotsProto::BallCollisionType::AVOID, {AutoChipOrKickMode::OFF, 0},
+            TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT, 0.0);
+        tactics_to_run[0].emplace_back(retreat_tactic);
+
+        event.common.set_tactics(tactics_to_run);
+    }
+}
+
+bool BallPlacementPlayFSM::shouldKickOffWall(const Update &event)
+{
+    // check if ball is too close to border
+    Point ball_pos        = event.common.world.ball().position();
+    Rectangle field_lines = event.common.world.field().fieldLines();
+
+    return !contains(field_lines, ball_pos);
+}
+
+bool BallPlacementPlayFSM::alignDone(const Update &event)
+{
+    std::optional<Robot> nearest_robot =
+        event.common.world.friendlyTeam().getNearestRobot(
+            event.common.world.ball().position());
+    if (nearest_robot.has_value())
+    {
+        return comparePoints(nearest_robot.value().position(), setup_point);
+    }
+    else
+    {
+        return false;
+    }
+}
+
+bool BallPlacementPlayFSM::kickDone(const Update &event)
+{
+    const auto ball_velocity = event.common.world.ball().velocity().length();
+    return (ball_velocity > SHOT_VELOCITY_THRESHOLD_M_PER_S) && !shouldKickOffWall(event);
+}
+
+bool BallPlacementPlayFSM::ballPlaced(const Update &event)
+{
+    Point ball_pos = event.common.world.ball().position();
+    std::optional<Point> placement_point =
+        event.common.world.gameState().getBallPlacementPoint();
+    std::vector<Robot> robots = event.common.world.friendlyTeam().getAllRobots();
+
+    // see if the ball is at the placement destination
+    if (placement_point.has_value())
+    {
+        return comparePoints(ball_pos, placement_point.value(), 0.05) &&
+               event.common.world.ball().velocity().length() < 0.1;
```

Comment:
so we have a "ball_not_kicked_threshold" defined in ShootOrPassPlayConfig (https://github.com/UBC-Thunderbots/Software/blob/8607d4565479f032a11333eca817026bca9b8873/src/proto/parameters.proto#L229)

Can we move that constant into a new config proto called AiParameterConfig and use that in ShootOrPassPlayFSM and BallPlacementPlayFSM?